pkg/runtime/grpc/grpc-runtime.go

```diff
@@ -0,0 +1,418 @@
+	}
+
+	results := make(map[string][]byte)
+	var resultsLock sync.Mutex
+
+	wg := sync.WaitGroup{}
+	for _, pod := range pods {
+		wg.Add(1)
+		go func(pod v1.Pod) {
+			gadgetCtx.Logger().Debugf("running gadget on node %q", pod.Spec.NodeName)
+			res, err := r.runGadget(gadgetCtx, pod)
+			if err != nil {
+				gadgetCtx.Logger().Errorf("node %q: %w", pod.Spec.NodeName, err)
+			}
+			if res != nil {
+				resultsLock.Lock()
+				results[pod.Spec.NodeName] = res
+				resultsLock.Unlock()
+			}
+			wg.Done()
+		}(pod)
+	}
+
+	wg.Wait()
+	return results, nil
+}
+
+func (r *Runtime) runGadget(gadgetCtx runtime.GadgetContext, pod v1.Pod) ([]byte, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dialOpt := grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
+		return NewK8SExecConn(ctx, pod, time.Second*30)
+		// return NewK8SPortForwardConn(ctx, s, time.Second*30)
+	})
+
+	conn, err := grpc.DialContext(ctx, "", dialOpt, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+	client := pb.NewGadgetManagerClient(conn)
+
+	allParams := make(map[string]string)
+	gadgetCtx.GadgetParams().CopyToMap(allParams, "")
+	gadgetCtx.OperatorsParamCollection().CopyToMap(allParams, "operator.")
+
+	runRequest := &pb.GadgetRunRequest{
+		GadgetName:     gadgetCtx.GadgetDesc().Name(),
+		GadgetCategory: gadgetCtx.GadgetDesc().Category(),
+		Params:         allParams,
+		Nodes:          nil,
+		FanOut:         false,
+		LogLevel:       uint32(gadgetCtx.Logger().GetLevel()),
+	}
+
+	runClient, err := client.RunGadget(ctx)
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return nil, err
+	}
+
+	controlRequest := &pb.GadgetControlRequest{Event: &pb.GadgetControlRequest_RunRequest{RunRequest: runRequest}}
+	err = runClient.Send(controlRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	parser := gadgetCtx.Parser()
+
+	jsonHandler := func([]byte) {}
+	jsonArrayHandler := func([]byte) {}
+
+	if parser != nil {
+		jsonHandler = parser.JSONHandlerFunc()
+		jsonArrayHandler = parser.JSONHandlerFuncArray(pod.Spec.NodeName)
+	}
+
+	doneChan := make(chan bool)
+
+	var result []byte
+	expectedSeq := uint32(1)
+
+	go func() {
+		for {
+			ev, err := runClient.Recv()
+			if err != nil {
+				break
+			}
+			switch ev.Type {
+			case pb.EventTypeGadgetPayload:
+				if len(ev.Payload) > 0 && ev.Payload[0] == '[' {
+					jsonArrayHandler(ev.Payload)
+					continue
+				}
+				if expectedSeq != ev.Seq {
+					gadgetCtx.Logger().Warnf("expected seq %d, %d messages dropped", expectedSeq, ev.Seq-expectedSeq)
+				}
+				expectedSeq = ev.Seq + 1
+				jsonHandler(ev.Payload)
+			case pb.EventTypeGadgetResult:
+				result = ev.Payload
+			case pb.EventTypeGadgetJobID:
+			// not needed right now
+			case pb.EventTypeGadgetDone:
+				gadgetCtx.Logger().Debug("got EventTypeGadgetDone from server")
+				doneChan <- true
+				return
```

Comment:
That would introduce labels, and I try to avoid them, if possible. Could use `defer func() { doneChan <- true }` at the top of the goroutine, though. Any preference?

Reply:
Let's keep your proposal.